### PR TITLE
Core Java Twitter Release

### PR DESCRIPTION
**Pull request by vassairm**

### CODE REVIEW

File: core_java/twitter/README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
+# Twitter CLI App
+## Introduction
+The Twitter CLI App provides a basic interface with which a user can interact with Twitter. 
+The application provides 3 functions, post a tweet with or without Geolocation, view a Tweet 
+(with only certain fields if desired), and delete one or more tweets.
+
+This project introduced the full MVC architecture, Apache web libraries, Spring and SpringBoot,
```
Comment (jarviscanada) [inline core_java/twitter/README.md]: Apache HttpComponents

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/controller/TwitterController.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.Arrays;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@org.springframework.stereotype.Controller
+public class TwitterController implements Controller {
+
+  private Service twitterService;
+
+  // Default constructor
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/controller/TwitterController.java]: remove unnecessary comments.
or use javadoc instead

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/controller/TwitterController.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  public TwitterController() {
+    twitterService = new TwitterService();
+  }
+
+  // Alternate constructor to provide your own TwitterService
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/controller/TwitterController.java]: use javadoc or remove comment

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/controller/TwitterController.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    this.twitterService = twitterService;
+  }
+
+  /**
+   * Parse user argument and post a tweet by calling service classes.
+   *
+   * @param args Post contents and optionally a longitude and latitude
+   * @return a posted tweet
+   * @throws IllegalArgumentException if args are invalid
+   */
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/controller/TwitterController.java:31]: good job. this is a proper java doc

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+  private static final String DELETE_URL = BASE_URL + "statuses/destroy/";
+
+  private static final int HTTP_OK = 200;
+
+  private HttpHelper httpHelper;
+  private Logger daoLogger = LoggerFactory.getLogger(TwitterDao.class);
+
+  /**
+   * Default Constructor, gets the OAuth values for TwitterHttpHelper itself.
+   */
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java]: remove

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        System.getenv("CONSUMER_SECRET"),
+        System.getenv("ACCESS_TOKEN"),
+        System.getenv("ACCESS_SECRET"));
+  }
+
+  /**
+   * Alternate constructor if you want to pass in your own TwitterHttpHelper.
+   *
+   * @param helper a TwitterHttpHelper
+   */
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java]: remove

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+  private static final String BASE_URL = "https://api.twitter.com/1.1/";
+  private static final String POST_URL = BASE_URL + "statuses/update.json";
+  private static final String SHOW_URL = BASE_URL + "statuses/show/";
+  private static final String DELETE_URL = BASE_URL + "statuses/destroy/";
+
+  private static final int HTTP_OK = 200;
+
+  private HttpHelper httpHelper;
+  private Logger daoLogger = LoggerFactory.getLogger(TwitterDao.class);
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java]: use IDE to rename daoLogger to logger (IDE > right-click > refactor )..
logger is a common naming convention

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+      urlParams = "?status=" + URLEncoder.encode(text, StandardCharsets.UTF_8.name());
+      if (location != null) {
+        float longitude = entity.getLocation().getCoordinates()[0];
+        float latitude = entity.getLocation().getCoordinates()[1];
+        urlParams += "&long="
+          + URLEncoder.encode(Float.toString(longitude), StandardCharsets.UTF_8.name())
+          + "&lat=" + URLEncoder.encode(Float.toString(latitude), StandardCharsets.UTF_8.name());
+      }
+    } catch (UnsupportedEncodingException ueex) {
+      daoLogger.error("Encoding not supported: " + StandardCharsets.UTF_8.name());
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java]: do you need to throw another exception here?

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/helper/TwitterHttpHelper.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+  private final String CONSUMER_KEY;
+  private final String CONSUMER_SECRET;
+  private final String ACCESS_KEY;
+  private final String ACCESS_SECRET;
+  private OAuthConsumer oac;
+  private HttpClient client;
+  private Logger httpLogger;
+
+  // Default Constructor
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/helper/TwitterHttpHelper.java]: go through your code base and remove all unnecessary comments.

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/helper/TwitterHttpHelper.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+  public TwitterHttpHelper() {
+    CONSUMER_KEY = System.getenv("CONSUMER_TOKEN");
+    CONSUMER_SECRET = System.getenv("CONSUMER_SECRET");
+    ACCESS_KEY = System.getenv("ACCESS_TOKEN");
+    ACCESS_SECRET = System.getenv("ACCESS_SECRET");
+    httpLogger = LoggerFactory.getLogger(TwitterHttpHelper.class);
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/helper/TwitterHttpHelper.java]: logger

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/helper/TwitterHttpHelper.java [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+   * @param uri Remote resource to access
+   * @return an HTTP Response for the given URI's request
+   */
+  @Override
+  public HttpResponse httpGet(URI uri) {
+    HttpGet get = new HttpGet(uri);
+    return getResponse(get);
+  }
+
+  private HttpResponse getResponse(HttpUriRequest request) throws RuntimeException {
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/helper/TwitterHttpHelper.java]: throwing runtime exception is not necessary.

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/service/TwitterService.java [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+  /**
+   * Search a tweet by ID.
+   *
+   * @param id     tweet id
+   * @param fields set fields not in the list to null
+   * @return Tweet object which is returned by the Twitter API
+   * @throws IllegalArgumentException if id or fields param is invalid
+   */
+  @Override
+  public Tweet showTweet(String id, String[] fields) {
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/service/TwitterService.java:94]: good job

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/service/TwitterService.java [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+      long idNum;
+      try {
+        idNum = Long.parseLong(id);
+      } catch (NumberFormatException nfex) {
+        throw new IllegalArgumentException(id + " is not a valid Tweet ID.");
+      }
+      try {
+        deletedTweets.add(dao.deleteById(idNum));
+      } catch (IllegalArgumentException ignored) {
+        // TwitterDao.deleteByID() throws IllegalArgumentException if we can't delete a tweet with
```
Comment (jarviscanada) [inline core_java/twitter/src/main/java/ca/jrvs/apps/twitter/service/TwitterService.java]: is this necessary?
if not, please remove and clean up.

**Review by jarviscanada — CHANGES_REQUESTED** [review]
good job. Very clean and readable code. 
please fix all comments.

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java (hunk 8 above)
Comment (vassairm) [reply core_java/twitter/src/main/java/ca/jrvs/apps/twitter/dao/TwitterDao.java]: I assumed UTF-8 would always be supported, so I didn't think to throw a new exception myself. Not entirely sure what kind of exception I would throw here, besides the same UnsupportedEncodingException, or maybe RuntimeException.

File: core_java/twitter/src/main/java/ca/jrvs/apps/twitter/service/TwitterService.java (hunk 13 above)
Comment (vassairm) [reply core_java/twitter/src/main/java/ca/jrvs/apps/twitter/service/TwitterService.java]: TwitterDAO throws IllegalArgumentException if a delete fails, but this has no bearing on whether we should continue deleting or not. As per Google Style Guide, if you catch and ignore an exception it must be named "ignored" and you must have a comment explaining why you ignore it. I may have been too wordy there though.

Comment (jarviscanada) [reply core_java/twitter/src/main/java/ca/jrvs/apps/twitter/service/TwitterService.java]: please mark it resolved.
